fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/ingredient/CustomIngredientImpl.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.recipe.ingredient;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.recipe.Ingredient;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.recipe.v1.ingredient.CustomIngredient;
+import net.fabricmc.fabric.api.recipe.v1.ingredient.CustomIngredientSerializer;
+import net.fabricmc.fabric.impl.recipe.ingredient.builtin.AndIngredient;
+import net.fabricmc.fabric.impl.recipe.ingredient.builtin.DifferenceIngredient;
+import net.fabricmc.fabric.impl.recipe.ingredient.builtin.NbtIngredient;
+import net.fabricmc.fabric.impl.recipe.ingredient.builtin.OrIngredient;
+
+public class CustomIngredientImpl extends Ingredient {
+	// Static helpers used by the API
+
+	public static final String TYPE_KEY = "fabric:type";
+	public static final int PACKET_MARKER = -1;
+
+	static final Map<Identifier, CustomIngredientSerializer<?>> REGISTERED_SERIALIZERS = new ConcurrentHashMap<>();
+
+	public static void registerSerializer(CustomIngredientSerializer<?> serializer) {
+		Objects.requireNonNull(serializer.getIdentifier(), "CustomIngredientSerializer identifier may not be null.");
+
+		if (REGISTERED_SERIALIZERS.putIfAbsent(serializer.getIdentifier(), serializer) != null) {
+			throw new IllegalArgumentException("CustomIngredientSerializer with identifier " + serializer.getIdentifier() + " already registered.");
+		}
+	}
+
+	@Nullable
+	public static CustomIngredientSerializer<?> getSerializer(Identifier identifier) {
+		Objects.requireNonNull(identifier, "Identifier may not be null.");
+
+		return REGISTERED_SERIALIZERS.get(identifier);
+	}
+
+	static {
+		CustomIngredientSerializer.register(AndIngredient.SERIALIZER);
```

Comment:
Question, is this OK class loading wise? Might be better moved to an init entrypoint to be sure its registered?

Reply:
It works fine because the various ingredients don't reference `CustomIngredientImpl` (as they're using only the API).

Reply:
Ah, this is fine. just seen that `CustomIngredientSerializer.get` actually calls `CustomIngredientImpl.getSerializer`. 👍 Might still be a tiny bit cleaner in the mod entrypoint, but not a major issue.